### PR TITLE
Bugfix in URI conversion when working with directories

### PR DESCRIPTION
**Pull request by malhotrashivam**
This PR fixes a bug in the conversion code for String -> URI when working with directories.
This bug could lead to failure in reading flat partitioned parquet files from S3 using a URI which doesn't end with a `\` character.

### CODE REVIEW

File: Base/src/test/java/io/deephaven/base/FileUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import junit.framework.TestCase;
+import org.junit.Assert;
+import org.junit.rules.TemporaryFolder;
+
+public class FileUtilsTest extends TestCase {
+
+    public void testConvertToFileURI() throws IOException {
+        final TemporaryFolder folder = new TemporaryFolder();
+        folder.create();
```
Comment (devinrsmith) [inline Base/src/test/java/io/deephaven/base/FileUtilsTest.java:19]: Seems odd to me that we actually want to create a temporary folder just so we can test URI stuff. There may be something valid about wanting this to work on Windows as well as Unix, but you could just do it based on the current directory; `Path.of("").toAbsolutePath().toString()`.

File: Base/src/test/java/io/deephaven/base/FileUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+        // Check if multiple slashes get normalized
+        fileUriTestHelper(someDir.getPath() + "////path///to////directory", true, expectedURI);
+
+        // Check if multiple slashes in the beginning get normalized
+        fileUriTestHelper("////" + someDir.getPath() + "/path/to/directory", true, expectedURI);
+    }
+
+    private static void fileUriTestHelper(final String filePath, final boolean isDirectory,
+            final String expctedURIString) {
```
Comment (devinrsmith) [inline Base/src/test/java/io/deephaven/base/FileUtilsTest.java:39]: misspelled.

File: Base/src/test/java/io/deephaven/base/FileUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        Assert.assertEquals(expctedURIString, FileUtils.convertToURI(filePath, isDirectory).toString());
+        Assert.assertEquals(expctedURIString, FileUtils.convertToURI(new File(filePath), isDirectory).toString());
+        Assert.assertEquals(expctedURIString, FileUtils.convertToURI(Path.of(filePath), isDirectory).toString());
+    }
+
+    public void testConvertToS3URI() throws URISyntaxException {
+        assertEquals("s3:/bucket/key", FileUtils.convertToURI("s3:/bucket/key", false).toString());
```
Comment (devinrsmith) [inline Base/src/test/java/io/deephaven/base/FileUtilsTest.java:46]: Hmmm; aren't s3 URIs supposed to look like `s3://bucket/some/key/path`? How do we currently handle `s3:/bucket/key`?

**Review by rcaudy — DISMISSED** [review]

File: Base/src/test/java/io/deephaven/base/FileUtilsTest.java (hunk 3 above)
Comment (malhotrashivam) [reply Base/src/test/java/io/deephaven/base/FileUtilsTest.java]: You are right, I did have a bad input and have fixed the test for now. 
Regarding `s3:/bucket/key`, will check and get back to you.

File: Base/src/test/java/io/deephaven/base/FileUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    private static void fileUriTestHelper(final String filePath, final boolean isDirectory,
+            final String expectedURIString) {
+        Assert.assertEquals(expectedURIString, FileUtils.convertToURI(filePath, isDirectory).toString());
+        Assert.assertEquals(expectedURIString, FileUtils.convertToURI(new File(filePath), isDirectory).toString());
+        Assert.assertEquals(expectedURIString, FileUtils.convertToURI(Path.of(filePath), isDirectory).toString());
+    }
+
+    public void testConvertToS3URI() throws URISyntaxException {
+        assertEquals("s3://bucket/key", FileUtils.convertToURI("s3://bucket/key", false).toString());
```
Comment (devinrsmith) [inline Base/src/test/java/io/deephaven/base/FileUtilsTest.java:42]: We should test `convertToURI("s3://bucket/key/", false)` and `convertToURI("s3://bucket/key//", false)`. I don't know what we want / expect in this situation. Similar Q for files.

Comment (malhotrashivam) [reply Base/src/test/java/io/deephaven/base/FileUtilsTest.java:42]: I think these would be bad inputs. I can maybe raise an exception in this case.

File: Base/src/test/java/io/deephaven/base/FileUtilsTest.java (hunk 3 above)
Comment (malhotrashivam) [reply Base/src/test/java/io/deephaven/base/FileUtilsTest.java]: The convertToURI code would return the same output for this input `"s3:/bucket/key/"`.

I tried fetching such a URI: `"s3:/dh-s3-parquet-test1/singleColFile.parquet"` and got the following error:
```
RuntimeError: java.lang.IllegalArgumentException: Invalid S3 URI: bucket not included: s3:/dh-s3-parquet-test1/singleColFile.parquet
	at software.amazon.awssdk.services.s3.S3Utilities.parseAwsCliStyleUri(S3Utilities.java:387)
	at software.amazon.awssdk.services.s3.S3Utilities.parseUri(S3Utilities.java:296)
	at io.deephaven.extensions.s3.S3SeekableChannelProvider.getReadChannel(S3SeekableChannelProvider.java:82)
	at io.deephaven.util.channel.CachedChannelProvider.getReadChannel(CachedChannelProvider.java:98)
	at io.deephaven.parquet.base.ParquetFileReader.<init>(ParquetFileReader.java:98)
	at io.deephaven.parquet.base.ParquetFileReader.create(ParquetFileReader.java:73)
```
So looks like it gets handled at the lower layer.

File: Base/src/test/java/io/deephaven/base/FileUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        fileUriTestHelper(currentDir.getPath() + "/path/to/directory", true, expectedURI);
+
+        // Check if multiple slashes get normalized
+        fileUriTestHelper(currentDir.getPath() + "////path///to////directory////", true, expectedURI);
+
+        // Check if multiple slashes in the beginning get normalized
+        fileUriTestHelper("////" + currentDir.getPath() + "/path/to/directory", true, expectedURI);
+
+        // Check for bad inputs
+        fileUriTestHelper(someFile.getPath() + "/", false, someFile.toURI().toString());
```
Comment (devinrsmith) [inline Base/src/test/java/io/deephaven/base/FileUtilsTest.java:34]: This is inconsistent with 

```java
fileUriTestHelper("file:" + someFile.getPath() + "/", false, someFile.toURI().toString());
```

which I _think_ we would either want to succeed or fail in the same way. Maybe we fix this by routing null *or* "file" to the `new File` path?

**Review by devinrsmith — CHANGES_REQUESTED** [review]

**Review by devinrsmith — APPROVED** [review]

**Pull request merged**